boards/esp32c3-devkit/include/periph_conf.h

```diff
@@ -0,0 +1,161 @@
+ *
+ * @note
+ * Most definitions can be overridden by an \ref esp32_application_specific_configurations
+ * "application-specific board configuration" if necessary.
+ *
+ * @file
+ * @author      Gunar Schorcht <gunar@schorcht.net>
+ */
+
+#ifndef PERIPH_CONF_H
+#define PERIPH_CONF_H
+
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @name    ADC and DAC channel configuration
+ * @{
+ */
+/**
+ * @brief   Declaration of GPIOs that can be used as ADC channels
+ *
+ * For generic boards, all ADC pins that have broken out are declared as ADC
+ * channels.
+ *
+ * @note As long as the GPIOs listed in ADC_GPIOS are not initialized as ADC
+ * channels with the `adc_init` function, they can be used for other
+ * purposes.
+ */
+#ifndef ADC_GPIOS
+#define ADC_GPIOS   { GPIO0, GPIO1, GPIO2, GPIO3, GPIO4, GPIO5 }
+#endif
+/** @} */
+
+/**
+ * @name   I2C configuration
+ *
+ * For generic boards, only one I2C interface I2C_DEV(0) is defined.
+ *
+ * The GPIOs listed in the configuration are only initialized as I2C signals
+ * when module `periph_i2c` is used. Otherwise they are not allocated and
+ * can be used for other purposes.
+ *
+ * @{
+ */
+#ifndef I2C0_SPEED
+#define I2C0_SPEED  I2C_SPEED_FAST  /**< I2C bus speed of I2C_DEV(0) */
+#endif
+#ifndef I2C0_SCL
+#define I2C0_SCL    GPIO4           /**< SCL signal of I2C_DEV(0) */
+#endif
+#ifndef I2C0_SDA
+#define I2C0_SDA    GPIO5           /**< SDA signal of I2C_DEV(0) */
+#endif
+/** @} */
+
+/**
+ * @name   PWM channel configuration
+ *
+ * For generic boards, two PWM devices are configured. These devices
+ * contain all GPIOs that are not defined as I2C, SPI or UART for this board.
+ * Generally, all outputs pins could be used as PWM channels.
+ *
+ * @note As long as the according PWM device is not initialized with
+ * the `pwm_init`, the GPIOs declared for this device can be used
+ * for other purposes.
+ *
+ * @{
+ */
+
+/**
+ * @brief Declaration of the channels for device PWM_DEV(0),
+ *        at maximum PWM_CHANNEL_NUM_DEV_MAX.
+ */
+#ifndef PWM0_GPIOS
+#define PWM0_GPIOS  { GPIO3, GPIO4 }
+#endif
+
+/** @} */
+
+/**
+ * @name    SPI configuration
+ *
+ * @note The GPIOs listed in the configuration are first initialized as SPI
+ * signals when the corresponding SPI interface is used for the first time
+ * by either calling the `spi_init_cs` function or the `spi_acquire`
+ * function. That is, they are not allocated as SPI signals before and can
+ * be used for other purposes as long as the SPI interface is not used.
+ * @{
+ */
+#ifndef SPI0_CTRL
+#define SPI0_CTRL   FSPI    /**< FSPI is used as SPI_DEV(0) */
+#endif
+#ifndef SPI0_SCK
+#define SPI0_SCK    GPIO6   /**< FSPI SCK (pin FSPICLK) */
+#endif
+#ifndef SPI0_MISO
+#define SPI0_MISO   GPIO2   /**< FSPI MISO (pin FSPIQ) */
+#endif
+#ifndef SPI0_MOSI
+#define SPI0_MOSI   GPIO7   /**< FSPI MOSI (pin FSPID) */
+#endif
+#ifndef SPI0_CS0
+#define SPI0_CS0    GPIO10  /**< FSPI CS0 (pin FSPICS0) */
+#endif
```

Comment:
Is this really better than

```C
static const spi_conf_t spi_config[] = {
    {
        .ctrl = FSPI,
        .sck = GPIO6,
        .mosi = GPIO7,
        .miso = GPIO2,
        .cs = GPIO10,
    },
};
```

Reply:
This is exactly the same way as done for all ESP32 boards. It gives the flexibility to use arbitrary GPIOs for the SPI interface without changing the board definition. This in turn allows to use an existing board definition for a slightly differing board.

Reply:
Hm I suppose you can mux any pin to SPI then?
If that brings you convenience for a use case, I'm not objecting it.
But generally boards should follow common RIOT conventions, CPU family specific conventions should be avoided if possible.

Reply:
> Hm I suppose you can mux any pin to SPI then?

Yes, although there are GPIOs with up to 6 "Direct I/O" functions using an IO_MUX to connect a peripheral signal to the pin (comparable to alternate functions of other MCUs), each peripheral signal can also be routed to any pin via the GPIO matrix. Whether the IO_MUX or the GPIO matrix is used is determined automatically when pins are initialized for the peripherals.

Of course, direct I/O via the IO_MUX is faster than using the GPIO matrix. For example, the GPIO matrix allows SPI signals with clock frequencies of "only" up to 40 MHz, in contrast to 80 MHz when using the IO_MUX. But this has no practical relevance, because the SPI clock rate may anyway only be a maximum of one fifth of the APB clock (normally 80 MHz).

This flexibility in defining the pins for the peripherals led to different users using different pins for the peripherals when using the same board, e.g. `esp32-wroom-32`. Therefore the ESP32x boards define only default values for the pins of the peripherals, which can be easily redefined at the command line or the Makefile of the application.

Another reason why I don't like to change it is that the rather comprehensive ESP32x SoC documentation is fully compliant with the approach used :wink:

> But generally boards should follow common RIOT conventions, CPU family specific conventions should be avoided if possible.

That is clear. But does it really make a difference for the user whether the GPIOs are set directly or via defines in the configuration?

Reply:
Consistency with other boards is a virtue, it's also easier when creating a new board definition that might e.g. have two SPI busses configured.